### ISSUE TITLE
[Bug]: `react/no-children-prop` does not report function as `children` in the child position when `allowFunctions` is `true`

### ISSUE THREAD

**Issue by jasongerbes-vista** [yannickcr/eslint-plugin-react#4005]
### Is there an existing issue for this?

- [x] I have searched the existing issues and my issue is unique
- [x] My issue appears in the command-line and not only in the text editor

### Description Overview

When `allowFunctions` is set to `true`, the `react/no-children-prop` rule does not report function as `children` in the child position, which contradicts the documented behaviour for this option.

According to the [rule docs](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-children-prop.md) under `allowFunctions`:

> When `true`, and passing a function as `children`, it must be in prop position and not child position.
>
> The following patterns are considered warnings:
>
> ```js
> <MyComponent>{data => data.value}</MyComponent>
> React.createElement(MyComponent, {}, data => data.value)
> ```

ES Lint config:

```js
"react/no-children-prop": ["error", {
  "allowFunctions": true
}]
```

Example:

```js
<MyComponent>{data => data.value}</MyComponent>
```

No error is reported for the above code.

This contradicts the documentation, which states that when `allowFunctions` is `true`, functions should not be passed as `children` in the child position.

### Expected Behavior

When `allowFunctions` is `true`, functions as `children` in the child position should be reported as incorrect code.

### eslint-plugin-react version

v7.37.5

### eslint version

v9.39.1

### node version

v24.12.0
